### ISSUE TITLE
Neural Network Invert Hash Readable Model doesn’t contain the full network parameters

### ISSUE THREAD

**Issue by apoorvreddy** [JohnLangford/vowpal_wabbit#1169]
### TL;DR:
The file generated using the —invert_hash flag doesn’t contain a few of the output_hidden weights (weights from the hidden layer to the output layer) and the output node bias, although the corresponding hashes for these edges exist in just the readable_model format (which only contains the hash:value tuples)  

#### Train Dataset
```
0 | price:.23 sqft:.25 age:.05 type=?
1 | price:.40 sqft:.35 age:.25 type=?
1 | price:.18 sqft:.15 age:.35 type=apartment
0 | price:.53 sqft:.32 age:.87 type=apartment
```

#### Train Command
`vw house_dataset.1 --nn 3 --holdout_off -f 1.model --readable_model 1.model.hash --invert_hash 1.model.human`

**Test Dataset**
```
1 | price:.18 sqft:.15 age:.35 type=apartment
```

`vw -i 1.model -t test -p test1.pred -r test1.pred.raw --audit > test1.log`

**Audit log**
```
0.150644
	Constant:202096:1:0.05359	type=apartment:256200:1:0.0420145	age:136516:0.35:0.067283	sqft:127124:0.15:0.116039	price:133176:0.18:0.0782487
0.190004
	Constant:202097:1:0.0728486	type=apartment:256201:1:0.0496264	age:136517:0.35:0.0781409	sqft:127125:0.15:0.152841	price:133177:0.18:0.0958506
-0.149247
	Constant:202098:1:-0.0528563	type=apartment:256202:1:-0.0416767	age:136518:0.35:-0.0670522	sqft:127126:0.15:-0.114896	price:133178:0.18:-0.0778418
-0.003480
	:156910:1:-0.00348006
0.099359
	:156911:1:0.0993585
0.007512
	:156912:1:0.00751181
0.410331
	:156913:1:0.39331	

:156911:0.187747:0.0993585	:156912:-0.14815:0.00751181	:156910:0.149512:-0.00348006
```

`x           = [0.150644, 0.190004, -0.149247]` are the input to the hidden nodes tanh function
`tanh(x)  = [0.149515, 0.18775, -0.148149]`

VW uses some numerical optimisations for tanh and so VW computes 
`tanh(x) = [0.149512, 0.187747, -0.14815]`

This means that the corresponding output_hidden weights that are being multiplied with these activation values are `[-0.00348006 , 0.0993585, 0.00751181]`.

Also `0.39331` is the bias in the output node. The hashes of the output_hidden weights are `[156910, 156911, 156912]` and the bias is `156913`

The issue is that the human-readable model (i.e. 1.model.human) generated using the invert_hash flag does not contain the following <hash, value> tuples:
`[156911 : 0.0993585, 156912 : 0.00751181, 156913 : 0.39331]` although these tuples exist in 1.model.hash (non hash_inverted readable model).

[vw_issue.zip](https://github.com/JohnLangford/vowpal_wabbit/files/715939/vw_issue.zip)

**Comment by apoorvreddy** [comment]
@arielf 

I noticed that there was a question related to this issue on stackoverflow
http://stackoverflow.com/questions/33441183/vowpal-wabbit-input-of-neural-network

Has there been any change to the code to print all the weights for --nn flag ?

**Comment by arielf** [comment]
I'm not aware of any intentional change to cause this.

Are you saying/hinting that there was a time in the past where this behaved differently?

The best way to make sure a feature you care about in `vw` never accidentally breaks is to add a test case for it in `test/RunTests`.  A failure of the test suite is pretty much a guarantee for blocking unintentional changes.

I'm not 100% sure, but I suspect that this behavior isn't new.

**Comment by apoorvreddy** [comment]
Thanks @arielf for replying ! :)

I should have phrased my question better.

This is what I meant to ask. Is there any way to print all the weights (and the bias) from the hidden layer to the output node, as they currently are not printed when using the invert_hash flag ? I believe this is an issue, as one cannot visualize the network without them.

**Comment by apoorvreddy** [comment]
The weights from the hidden layer to the output layer (hidden_output) has an empty dat.ns_pre, hence the human readable value of these weights are the same which is `[x]`, where x is the number of hidden nodes. (i.e. running with `-nn 3` gives `[3]`)

Because of this, `dat.all.name_index_map` (which is a std::map) stores only one of the hidden_output weights. There is a check at [gd.cc Line238](https://github.com/JohnLangford/vowpal_wabbit/blob/d5caa6e944a38f61c673f47027206eb8e9e7f1d7/vowpalwabbit/gd.cc#L238) which checks if the human readable value has already been saved.

```C++
    if (!dat.all.name_index_map.count(ns_pre))
	dat.all.name_index_map.insert(std::map< std::string, size_t>::value_type(ns_pre, index >> stride_shift));
```

When I tried adding the hash value to the ns_pre at [gd.cc Line 213](https://github.com/JohnLangford/vowpal_wabbit/blob/d5caa6e944a38f61c673f47027206eb8e9e7f1d7/vowpalwabbit/gd.cc#L213):

```C++
  if (ns_pre == "")
    ns_pre += "{" + std::to_string(index >> stride_shift) + "}";
```
The hidden_output weights are getting printed in the inverted_hash readable model as well. Now the human readable name file is given as: `{<hash value>}[3]:<hash value>:<weight>` for `-nn 3`

Can you suggest an appropriate place to update the dat.ns_pre variable as the above seems hacky to me ?

**Comment by arielf** [comment]
Hi @apoorvreddy ,

I'm not sure what's the best way to make this behave better. As you noted the code is pretty complex.  `--invert_hash` was added as an early reduction, and may not interface nicely with the core SGD code. `ns` stands for "name space", not sure about the meaning of `pre` in this context.

Trying to be practical: does the change above still pass the full `make test`?  If yes, I think this is a reasonable change and would encourage you to submit a PR after:

- Getting John's blessing for this (e.g. I prefer `[ ]` over the `{ }` notation here. It seems more consistent with what's printed by `--audit`)
- Adding a comment with this code so the next person understands why this is there
- Adding a test case for it in `test/RunTests` so if it breaks in the future, the problem will be caught immediately.

**Comment by AbdealiJK** [comment]
@arielf @JohnLangford I believe a better approach would be to actually label what the hash means manually - like how it's done for `Constant`.
I've made a PR at https://github.com/JohnLangford/vowpal_wabbit/pull/1179 which does this.

@apoorvreddy can you test if this works for you ?

**Comment by apoorvreddy** [comment]
@AbdealiJK This works for me 👍 Perhaps, add comments and tests as well ?
